### PR TITLE
Make the event-bus more robust and production ready

### PR DESCRIPTION
**Pull request by petereast**
Work done by @petereast - this forms a good base line to now iterate on so suggesting that this goes in as is - even though there is more work to be done on this library.

### CODE REVIEW

**Comment by diversemix** [comment]
My thoughts are that we should now get this in - to form a base line to work from? Thoughts?
/cc @erezmus @nlisgo ?

**Comment by erezmus** [comment]
Fine with me

**Comment by diversemix** [comment]
Going to hang fire on the merge until after the demo this afternoon!

**Comment by diversemix** [comment]
Ok, I have realized this is failing because its lacking tests for the `amqp-connector` - doing those now

File: lib/event-bus/src/rabbit-event-bus/index.test.ts [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+      setTimeout(() => {
+        expect(then).toHaveBeenCalledTimes(3);
+        done();
+      }, 250);
+    });
+  });
+
+  // Tests that it queues up messages that it can't send and that resends then when it can
+  // If a message fails to send it will retry it
```
Comment (nlisgo) [inline lib/event-bus/src/rabbit-event-bus/index.test.ts]: Shall we add these tests now or in a followup?

Comment (diversemix) [reply lib/event-bus/src/rabbit-event-bus/index.test.ts]: Ah - if you get an approval from me with comments then I'm happy it go in now or later - happy its up to you.

File: message-bus-tester/src/main.ts [hunk 2]
```diff
@@ -57,8 +57,9 @@ const init_mq = async () => {
       ...test_event_def
     }
 
-    rabbitmq_mq.publish(event);
-  }, 5000)
+    console.log("send attempt");
+    rabbitmq_mq.publish(event).then(() => console.log("send complete"));
+  }, 500)
```
Comment (nlisgo) [inline message-bus-tester/src/main.ts]: These changes were made here: https://github.com/libero/reviewer/commit/c978656c94bb66f8f5e2b172e2fe7a51b97eab95

Not sure the `console.log` is necessary and if not should be removed. @petereast can help us but maybe someone else can comment?

Comment (diversemix) [reply message-bus-tester/src/main.ts]: I'd expect that was debug left in by @petereast :grin:

File: lib/event-bus/src/rabbit-event-bus/amqp-connector.test.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            const connector = new AMQPConnector<{}>(url, channel(), [], [], 'service');
+            const handler = jest.fn().mockImplementation(async () => Promise.resolve(true));
+
+            await flushPromises();
+            await connector.subscribe({ kind: 'foo', namespace: 'bar' }, handler);
+
+            await flushPromises();
+            expect(handler).toHaveBeenCalledTimes(1);
+            expect(handler).toHaveBeenCalledWith('foo');
+            expect(mockChannel.ack).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (nlisgo) [reply lib/event-bus/src/rabbit-event-bus/amqp-connector.test.ts]: I've pushed up a commit which should address this but the comparison fails. Can't spot why yet. Just getting this message:

```
- Expected
+ Received

- {"content": {"toString": [Function toString]}},
+ {"content": {"toString": [Function toString]}},
```

File: message-bus-tester/src/main.ts (hunk 2 above)
Comment (petereast) [reply message-bus-tester/src/main.ts]: Yep, that's debug. As a rule of thumb, `console.*` should be removed, but `logger.*` should be left in. Sometimes you might need to lower the level of the logger statement, e.g. `logger.info` -> `logger.debug` or `logger.debug` -> `logger.trace`.

File: lib/event-bus/src/rabbit-event-bus/amqp-connector.test.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+            const connector = new AMQPConnector<{}>(url, channel(), [], [], 'service');
+            const handler = jest.fn().mockImplementation(async () => Promise.resolve());
+
+            await flushPromises();
+            await connector.subscribe({ kind: 'foo', namespace: 'bar' }, handler);
+
+            await flushPromises();
+            expect(handler).toHaveBeenCalledTimes(1);
+            expect(logger.warn).toHaveBeenCalledTimes(1);
+            expect(logger.warn).toHaveBeenCalledWith('eventHandlerFailure');
```
Comment (petereast) [inline lib/event-bus/src/rabbit-event-bus/amqp-connector.test.ts]: Worth keeping in mind for the future: standardising on log message format will make things a _*heck*_ of a lot easier later on when you need to trace across a few services.

File: lib/event-bus/src/rabbit-event-bus/amqp-connector.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nlisgo) [reply lib/event-bus/src/rabbit-event-bus/amqp-connector.test.ts]: Have fixed now. The problem was that it isn't possible to compare functions. They will never be equal to each other so we have to store them.

See: https://medium.com/@boriscoder/the-hidden-power-of-jest-matchers-f3d86d8101b0

[the remark this answers is not in the document]
Comment (nlisgo) [reply lib/event-bus/src/rabbit-event-bus/amqp-connector.test.ts]: Can we resolve this now?

File: message-bus-tester/src/main.ts (hunk 2 above)
Comment (nlisgo) [reply message-bus-tester/src/main.ts]: I'll remove.

**Pull request merged**